packages/yew/src/html/html_ref.rs

```diff
@@ -0,0 +1,391 @@
+/// }
+///
+/// pub struct Controller {
+///     log_ref: ComponentRef<MessageHolder>,
+/// }
+///
+/// impl Component for Controller {
+///     type Message = ();
+///     type Properties = ();
+///
+///     fn create(_ctx: &Context<Self>) -> Self {
+///         Self {
+///             log_ref: HtmlRef::default(),
+///         }
+///     }
+///
+///     fn view(&self, _ctx: &Context<Self>) -> Html {
+///         let onclick = {
+///             let log_ref = self.log_ref.clone();
+///             Callback::from(move |_| {
+///                 log_ref
+///                     .get()
+///                     .expect("a message holder")
+///                     .send_message("example message".to_string())
+///             })
+///         };
+///         html! {
+///             <>
+///                 <MessageHolder ref={&self.log_ref} />
+///                 <button {onclick}>{"Send example message"}</button>
+///             </>
+///         }
+///     }
+/// }
+/// ```
+/// ## Relevant examples
+/// - [`nested_list`](https://github.com/yewstack/yew/tree/master/examples/nested_list)
+pub struct HtmlRef<T: ErasedStorage> {
+    inner: Rc<dyn ErasedRef>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: ErasedStorage> fmt::Debug for HtmlRef<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("HtmlRef").finish_non_exhaustive()
+    }
+}
+
+impl<T: ErasedStorage> Clone for HtmlRef<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T: ErasedStorage> Default for HtmlRef<T> {
+    fn default() -> Self {
+        let inner: Rc<RefState<T::Erased>> = Rc::new(RefState {
+            binding: RefCell::new(None),
+        });
+        Self {
+            inner,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T: ErasedStorage> PartialEq for HtmlRef<T> {
+    fn eq(&self, othr: &Self) -> bool {
+        erased_eq(&self.inner, &othr.inner)
+    }
+}
+
+impl<T: ErasedStorage> HtmlRef<T> {
+    /// Create a new, unbound HtmlRef
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the referenced value, if it is bound
+    pub fn get_ref(&self) -> Option<impl '_ + Deref<Target = T>> {
+        let erased_ref = get_erased_ref::<T::Erased>(&self.inner)?;
+        Some(Ref::map(erased_ref, T::downcast_ref))
+    }
+
+    /// Get the referenced value, if the HtmlRef is bound
+    pub fn get(&self) -> Option<T>
+    where
+        T: Clone,
+    {
+        Some(self.get_ref()?.clone())
+    }
+
+    #[cfg(any(feature = "csr", feature = "ssr"))]
+    pub(crate) fn to_erased(&self) -> ErasedHtmlRef {
+        Some(self.clone()).into()
+    }
+}
+
+impl NodeRef {
+    /// Try converting the node reference into another form
+    pub fn cast<INTO: AsRef<Node> + From<JsValue>>(&self) -> Option<INTO> {
+        let node = self.get();
+        node.map(Into::into).map(INTO::from)
+    }
+}
```

Comment:
Nit: can you format it in a way that `impl` blocks come _after_ the definition of the type? It makes it easier to read and understand where the type is coming from